core/src/main/c/linux/filewatcher.c

```diff
@@ -0,0 +1,179 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific languageext install ms-vscode.cpptools governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+#include <jni.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <poll.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/inotify.h>
+#include <string.h>
+#include <unistd.h>
+#include <libgen.h>
+
+struct file_watcher
+{
+    // fd is the inotify api file descriptor
+    uintptr_t fd;
+    // wd is the watch descriptor for poll calls
+    int *wd;
+    // filename is the name of the watched file
+    char *filename;
+};
+
+/* Creates a file_watcher that uses inotify to watch a single file for changes
+
+   Args:
+   fp (filepath) is the memory address of string that holds the path of the file to be watched
+
+   Returns:
+   the address of the initialized file_watcher struct. If value is 0, then an error has occurred
+
+*/
+static uintptr_t setup(const char *fp)
+{
+    struct file_watcher *fw;
+
+    /* Create the file descriptor for accessing the inotify API. */
+    const int fd = inotify_init();
+    if (fd == -1)
+    {
+        return 0;
+    }
+
+    /* Allocate memory for watch descriptor. Return 0 if
+       there is a malloc error. */
+    int *wd = malloc(sizeof(int));
+    if (wd == NULL)
+    {
+        return 0;
+    }
+
+    /* Create a copy of the path so we can run dirname/basename on it. We use
+       inotify to watch the entire directory (and its contents) for changes
+       but only surface changes made to the watched file */
+    char pathCpy[strlen(fp)];
+    strcpy(pathCpy, fp);
```

Comment:
we should not need a copy of string

Reply:
I found that when I didn't copy, the `dirname` call attempted to modify the `fp` string, which segfaulted because it is in protected memory.  Per man pages:

> Both dirname() and basename() may modify the contents of path, so it may be desirable to pass a copy when calling one of these functions. 
https://linux.die.net/man/3/basename

I did find an extra, unnecessary `strcpy` which I deleted in 1264dacad